File: random/recursive_path_finding/python/recursivePathFinding.py
#This represents up, down left and right. Saves us having to enter co-ords when walking
directions = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1]
]


#maze is the maze string 2d array
#wall - define what character represents a wall
#curr is the current co-ords
#end is if we are at the end
#seen is a 2d array originally populated with falses but as we traverse gets populated with truthes
#path is the path we are currently on
def walk(maze, wall, curr, end, seen, path):

    #Base cases
    #1. Off the map
    if(curr['x'] < 0 or curr['x'] >= len(maze[0]) 
       or curr['y'] < 0 or curr['y'] >= len(maze)):
        return False

    #2. On a wall
    if(maze[curr['y']][curr['x']] == wall):
        return False
    
    #4. At the end
    if(curr['x'] == end['x'] and curr['y'] == end['y']):
        path.append(end)
        return True
    
    #3. Already seen this block
    if(seen[curr['y']][curr['x']]):
        return False
    
    seen[curr['y']][curr['x']] = True
    
    path.append(curr)
    
    #recursive step
    for direction in directions:
        x, y =  direction[1], direction[0]
        if walk(maze, 
             wall, 
             {'x': curr['x'] + x, 'y': curr['y'] + y},
             end, seen, path):
            return True

if __name__ == "__main__":
    print("Traversal Starting")
    maze = [
        ["#","#","#", "#", " ", "#", "#"],
        ["#"," "," ", " ", " ", "#", "#"],
        ["#"," ","#", "#", "#", "#", "#"]
    ]
    seen = [[False for x in range(len(maze[0]))] for y in range(len(maze))]
    path = []

    walk(maze, '#', {'x': 4, 'y': 0 }, {'x': 1, 'y': 2}, seen, path)
    print(path)

File: random/recursive_path_finding/java/recursivePathFinding.java
import java.util.ArrayList;

class recursivePathFinding {
    static int[][] directions = {
        {-1, 0}, //up
        {0, 1}, // right
        {1, 0}, //down
        {0, -1} //left
    };

    //5 parameters
    //maze - 2d string array representing maze
    //wall - character representing a wall
    //curr - the current co-ords
    //end - defines the end of the maze in co-ords
    //seen - 2d array representing maps but with True and False
    //path - current path the recursive algorithm has traversed
    public static boolean walk(String[][] maze, String wall, Point curr,
    Point end, Boolean[][] seen, ArrayList<Point> path)
    {
        //Base Cases
        //1. out of bounds
        if(curr.x < 0 || curr.x >= maze[0].length
        || curr.y < 0 || curr.y >= maze.length)
        {
            return false;
        }

        //2. hit a wall
        if(maze[curr.y][curr.x] == wall)
        {
            return false;
        }

        //3. already seen this block
        if(seen[curr.y][curr.x] == true)
        {
            return false;
        }

        //4. it is the end
        if(curr.x == end.x && curr.y == end.y)
        {
            path.add(curr);
            return true;
        }

        //adding this block to the seen blocks so we don't look over it again
        seen[curr.y][curr.x] = true;

        path.add(curr);
        //walk
        for(int i = 0; i < directions.length; i++)
        {
            Point next_curr = new Point(curr.x + directions[i][0], curr.y + directions[i][1]);
            Boolean succ = walk(maze, "X", next_curr, end, seen, path);
            if(succ == true)
            {
                return true;
            }
        }
        return false;
    }


    public static void printPath(ArrayList<Point> path)
    {
        for(int i = 0; i < path.size(); i++)
        {
            path.get(i).print();
        }
    }
    public static void main(String[] args)
    {
        String[][] maze = {
            {"X", "X", "X", "X", " ", "X"},
            {"X", "X", "X", "X", " ", "X"},
            {"X", "X", " ", " ", " ", "X"},
            {"X", "X", " ", "X", "X", "X"}
        };
        Point start = new Point(4, 0);
        Point end = new Point(2, 3);
        Boolean[][] seen = new Boolean[maze.length][maze[0].length];
        
        //populating the seen array with falses
        for(int i = 0; i < seen.length; i++)
        {
            for(int j = 0; j < seen[0].length; j++)
            {
                seen[i][j] = false;
            }
        }
        ArrayList<Point> path = new ArrayList<Point>();
        walk(maze, "X", start, end, seen, path);
        printPath(path);

    }
}
